src/context.rs

```diff
@@ -0,0 +1,48 @@
+use std::{
+    cell::{Cell, UnsafeCell},
+    marker::PhantomData,
+};
+
+use crate::cell::{GCell, Token};
+
+pub struct TraverseCtx<'a, 't> {
+    token: UnsafeCell<&'a mut Token<'t>>,
+
+    /// It is only defined to make sure the `TraverseCtx` is `!Sync`. Since negative traits are nighly.
+    /// We basically want our type to behave similar to the `Cell` since we both wrap `UnsafeCell`,
+    /// In a similar manner.
+    _cell_marker: PhantomData<Cell<&'a mut Token<'t>>>,
+}
+
+impl<'a, 't> TraverseCtx<'a, 't> {
+    pub fn new(token: &'a mut Token<'t>) -> Self {
+        Self {
+            token: UnsafeCell::new(token),
+            _cell_marker: PhantomData {},
+        }
+    }
+
+    pub fn get_node<'b, T>(&self, node_ref: &'b GCell<'t, T>) -> &'b T
+    where
+        'a: 'b,
+    {
+        // SAFETY: This can cause data races if called from a separate thread,
+        // but `TraverseCtx` is `!Sync` so this won't happen.
+        let tk = unsafe { &*self.token.get() };
```

Comment:
You are right thanks for the thorough explanation now I understand why it doesn't work, GhostCell only allows using the borrow without having a token to borrow something else from the tree. That's why having `borrow` and `borrow_mut` is necessary + using `clone` for children.

So I have to add some sort of runtime check which defeats the whole purpose of the `GhostCell`s and `GhostToken`s.

I'm going to close this PR.

I think this approach is the best we can do for the performance, If having a Babel-like API is the goal I believe we have to go for the use of indices and introduce indirection since it is really similar to how path works.